### DISCUSSION TITLE
Roadmap for Wegue v2

### DISCUSSION THREAD

**Discussion by chrismayer** [meggsimum/wegue#327] [Announcements]
This discussion should act as a kind of roadmap to collect things, which have to be done on the way towards Wegue v2.

### TODOs before merging #314 and #320

- [x] Merge #325 
- [x] Create `v1` branch -> https://github.com/wegue-oss/wegue/tree/v1
- [x] Known patches for #314
- [x] Modify constants inside unit tests for #320

### v2 must-have

- [x] #314 
- [x] #320
- [x] Upgrade tooling (dev-) dependencies, e.g. by #295, #319

### v2 nice-to-have

- [x] #329 
- [x] #330

**Comment by sronveaux** [comment]
Hi @chrismayer,

Simple question, where would you place #295 in this roadmap?
Having `Vue2 LTR` would be a big plus I think. However, whether this should be done before releasing the first v2 version is up to you...
Same for `Vuetify`, but LTR is not out yet and planned `Q2 2023` as stated on their site. So I suppose where it should be included will depend in their timing and `Wegue V2` timing too...

**Reply by chrismayer** [reply]
Not sure about #295. Maybe it is somehow superseded by #314 and #320? I think it should at least be re-validated after the mentioned PRs are merged.

**Reply by sronveaux** [reply]
I personally don't think it is superseded and even see it the other way around, after #314 will be merged and all the tool suite will be upgraded, putting up to date versions of the core components with #320 and #295 will take even more sense.
I don't know what others think about it but I really think it is an important part of this roadmap.

**Reply by chrismayer** [reply]
"Superseded" was meant in a way that it might be better to upgrade tooling according to the newly updated setup (after #314). I am sure more and newer upgrades are possible then. So I think starting this from scratch might be more straight-forward. So I agree that having an up-to-date tooling is an important part of the roadmap. But as said I am unsure, if #295 is the right way. Anyhow I add a general point to the roadmap above.

I'd see it as "must-have" since it might reduces more vulnerabilities. Any objections?

**Reply by sronveaux** [reply]
O.K., I understand what you meant here and totally agree with you on this one!

**Comment by fschmenger** [comment]
I will go through the issues and make some proposals in the upcoming weeks.

The general question is, if we should "release early and often" (as pointed out by @chrismayer in #321), which would get my vote.
Or should we rather take more time and incorporate more content.

Generally, I would propose the following structure:
* v2 must-have
* v2 nice-to-have
* later versions


So depending on the time horizon I`d rather put the OL 7.x ugprade into a later step (nice-to have or later), as the migration is probably not straightforward. 
I would also kindly ask you to not put loose content in here, which is not tracked inside an issue or PR.

**Comment by chrismayer** [comment]
I adapted the roadmap according to your feedback - thanks!. As said, this was only a kickstart collecting some things from other PRs / issue / discussions. Feel free to edit.

In general I also vote for "release early and often". So I put in the must-have / nice-to-have structure. Please have a look and review / edit accordingly.

@fschmenger For software development things I agree, that there should be issues and/or pull requests but for mini reminders like "Create `v1` branch"  creating an issue is too much overhead IMHO. But if you see a benefit / necessity in having a separate for every tiny step, feel free to create and manage those issues.

**Reply by fschmenger** [reply]
Hi Christian, thanks for updating the roadmap accordingly and wrapping some of the more important stuff by issues. Since branching away is an inheritent step of the roadmap, I agree, we don`t need an issue for this.

**Comment by spwoodcock** [comment]
**What is done**
- All of the TODO list appear to be done now, apart from creating the V1 branch.
- I just updated #320.
I created the commit to be built on top of #314, so you should only have to merge the single PR `#320` (it will include all commits from previous upgrades sequentially - hopefully a time saver).

**What is remaining**
- Point 3 on the **v2 must-have** list: `Upgrade tooling (dev-) dependencies, e.g. by #295, #319`.
- Creating the V1 branch.

**What I will be working on**
- I will work on top of PR #320 today to start on #329 and #330.

**Reply by spwoodcock** [reply]
I completed issue #330 and closed PR #320 in favour of PR #333.

**Comment by spwoodcock** [comment]
All tasks in the description should be done now with latest PR #335.

I stacked all PRs on top of one another, to allow for easier merging and testing of the final result.

The only thing missing is some thorough testing of the updates (outside of the basic app-starter template).

I will be working to add more OpenLayers functionality on top, but won't make any more PRs for now.

> This discussion should act as a kind of roadmap to collect things, which have to be done on the way towards Wegue v2.
> ### TODOs before merging #314 and #320
> 
>     * [x]  Merge [Maintenance v1.2.0 #325](https://github.com/wegue-oss/wegue/pull/325)
> 
>     * [ ]  Create `v1` branch
> 
>     * [x]  Known patches for [Update to last Vue CLI v5 template #314](https://github.com/wegue-oss/wegue/pull/314)
> 
>     * [x]  Modify constants inside unit tests for [Merge Vue CLI updates & upgrade to OpenLayers v6.15.7 #320](https://github.com/wegue-oss/wegue/pull/320)
> 
> 
> ### v2 must-have
> 
>     * [x]  [Update to last Vue CLI v5 template #314](https://github.com/wegue-oss/wegue/pull/314)
> 
>     * [x]  [Merge Vue CLI updates & upgrade to OpenLayers v6.15.7 #320](https://github.com/wegue-oss/wegue/pull/320)
> 
>     * [x]  Upgrade tooling (dev-) dependencies, e.g. by [Upgrade non-dev dependencies #295](https://github.com/wegue-oss/wegue/pull/295), [Upgrade MDI libraries #319](https://github.com/wegue-oss/wegue/pull/319)
> 
> 
> ### v2 nice-to-have
> 
>     * [x]  [Upgrade to latest OpenLayers #329](https://github.com/wegue-oss/wegue/issues/329)
> 
>     * [x]  [Upgrade to Vue 2 and Vuetify 2 LTS   #330](https://github.com/wegue-oss/wegue/issues/330)

**Comment by chrismayer** [comment]
Thanks @spwoodcock for your update and your ongoing tireless work! Nice to see, that mostly everything is prepared now. I opened a PR #336 as preparation for a an upcoming patch release. After #336 is reviewed and merged I'll do the `v1.2.1` release, which will be then the base for a `v1` branch as in the initial description of this discussion. Afterwards we are able to merge the stacked PR into the master branch.

**Comment by chrismayer** [comment]
I just released Wegue [v1.2.1](https://github.com/wegue-oss/wegue/releases/tag/v1.2.1) with the current master branch and created a [v1 branch](https://github.com/wegue-oss/wegue/tree/v1) right after the release. So now we have the pre-conditions to merge in the v2 work, which has been done in several PRs. I'd vote to merge in soon in order to get things a bit more streamlined. I see 2 ways of doing that

- Merge #314 and #320 (needs a conflict resolve) and merge upcoming work in #335 separately since there are failing tests and @fschmenger wished to give this a review or
- wait until everything is done in #335 and merge it 

I personally would merge #314 and #320 to have a working base with the new setup and decrease the pressure of having #335 to be fixed that quick. 

Any opinions on the precedure?

**Reply by fschmenger** [reply]
As far as I can recall #314 and #320 are good to go. So fine by me to merge those first.
Could you give it a quick review again when the conflicts are resolved, so nothing got accidently changed? Should be fairly easy, since only a very few things should be modified. (For that matter, maybe it's also worth rebasing before merge, so it's a bit simpler to see whats going on)

**Reply by fschmenger** [reply]
Sorry, just merge #314, it's fine. Rest above applies to #320 only.

**Reply by sronveaux** [reply]
Hi @chrismayer,

So nice to see v2 of Wegue becoming reality! Thanks for all the work done on this!

100% agree with you on this one, I personally think merging what's available can give the opportunity to have a working base with the new setup and also give the possibility to continue the development of new things expected for v2 in parallel without, as you said, push pressure on people involved in #335...

**Reply by chrismayer** [reply]
Thanks for your feedback @fschmenger and @sronveaux. OK, then I am going to merge #314 and #320 once the conflict is solved (I created a PR for that since I have (of course) no write access to the base branch, see https://github.com/EnviDat/wegue/pull/1

I'll update here once it is done.

**Reply by spwoodcock** [reply]
I cherry-picked the latest commit for v1.2.1 into #335. It's now ready to merge.

The easiest option would be to just merge #335 and close the rest (it includes all stacked commits).

However if it is preferred that they are merged sequentially, then:
- Thanks for @chrismayer #320 is mergable now.
- I resolved all the remaining merge conflicts on PRs I have made.

Good luck! Looking forward to Wegue V2 🎉

**Comment by chrismayer** [comment]
Since it was quite quiet here for a longer time, I just merged #320 to accelerate things up again. Preparing work by branching away the latest `v1` stage was already done, so I just added some info to the README that the scope of the master branch changed, see #337 (needs reviews :wink: )

It would be really cool now if the planned and missing reviews for #335 could be done soon, so this first major steps towards v2 could be seen as done. Thanks in advance :+1:

**Comment by fschmenger** [comment]
Hi guys,
I just wanted to let you know that I'm currently very limited on time that I can dedicate to Wegue. This will hopefully change by beginning of October. I`m available for questions but I cannot do longer integration tests and reviews as for example required in #329.
Sorry for that,
Felix

**Reply by spwoodcock** [reply]
Thanks @fschmenger, I'll probably be working a good amount at adding new features in the coming months. I'll keep working on my fork and PR everything that I can for review in the future 👍

**Comment by sronveaux** [comment]
Hi @chrismayer,

I'm still planning to add a little contribution before `Wegue V2` is out, which I think should be included. This is simply adding some update tips for the upgrade of `Vue-CLI` (and `Webpack` under the hood) inside `upgrade-notes.md` and a simple clean of unit tests as merging #335 reintroduced some tests inside a `test` directory which is not used anymore (and should be moved to `tests` or simply removed if already included there).

If I have enough time, I'll also tackle #322 but I'd prefer the problem of updated packages breaking unit tests to be resolved beforehand... I'll give a fast look as soon as I have time but hope you (or someone else in the team) has also noticed it and has some precise ideas about this...

Best regards,
Sébastien

**Reply by chrismayer** [reply]
Hi @sronveaux, the proposed contributions are very much appreciated. Looking forward to your input. Give me a ping if you need some reviews.

**Comment by chrismayer** [comment]
It's been a while and a lot of great work has been done. The TODO list of this roadmap is completed and even more has been done.

So I just released a first beta version of Wegue 2.0.0 https://github.com/wegue-oss/wegue/releases/tag/v2.0.0-beta.1, which can be used for some final testing. Since we tested mainly against the current master (which is the base of the release) and some of the core devs are using the master branch for productive Wegue apps there should not be huge surprises anymore. Therefore the plan is to have a small beta phase of 1 week up to 10 days, with hopefully less new beta versions. And then provide Wegue 2.0.0 with some more release notes...

Thanks already to all who pushed this forward :muscle: :+1:

**Comment by fschmenger** [comment]
Hi Christian,
great to see that we are on the verge of releasing v2.0!

Should we look into #365 which emerged in the process of all the upgrades or do you consider it a minor bug? I proposed a solution within #368 but have currently no time to work on it.

Otherwise thanks to everyone for all the efforts from my part.
I hopefully can dedicate some time to the project again by autumn.

Cheers Felix

**Comment by chrismayer** [comment]
Hi Felix,
thanks for your feedback. Of course it would be really cool, to have #365 fixed ASAP, but since there are no resources to fix this in the near future I'd suggest to go for 2.0.0 with this as a known issue. Otherwise the release will be pushed back even further, so I have the fear to loose the earnings of the work since we are chasing behind the base libs all the time.

**Comment by sronveaux** [comment]
Hi team,

That's good news ! Having V2.0.0 out soon is a good idea I think as we have something quite updated and stable here (except for Vue and Vuetify which is another story...)

As already stated, the roadmap was finished a few months ago now and I also think postponing it is taking a risk to see it coming out outdated which would be sad when we see all the good work that the whole team made on this. We should never forget that software is never finished, only released !  
For sure #365 is a known issue which would be great if fixed shortly and will worth a V2.1.0 for itself !

Thanks to all contributors by the way for the very interesting colaborative work made on this great framework !

**Comment by chrismayer** [comment]
# Finally!

Version 2.0.0 is there: https://github.com/wegue-oss/wegue/releases/tag/v2.0.0

Also here many thanks to all who contributed in order to make 2.0.0 real and thanks for the good input/discussions bringing the project really forward :tada:
